modAion/src/org/aion/zero/db/AionContractDetailsImpl.java

```diff
@@ -79,26 +81,50 @@ public AionContractDetailsImpl(byte[] code) throws Exception {
         decode(code);
     }
 
+    /**
+     * Adds the key-value pair to the database unless value is an IDataWord whose underlying byte
+     * array consists only of zeros. In this case, if key already exists in the database it will be
+     * deleted.
+     *
+     * @param key The key.
+     * @param value The value.
+     */
     @Override
-    public void put(DataWord key, DataWord value) {
-        if (value.equals(DataWord.ZERO)) {
+    public void put(IDataWord key, IDataWord value) {
+        // We strip leading zeros of a DataWord but not a DoubleDataWord so that when we call get
+        // we can differentiate between the two.
+
+        if (value.isZero()) {
             storageTrie.delete(key.getData());
         } else {
-            storageTrie.update(key.getData(), RLP.encodeElement(value.getNoLeadZeroesData()));
+            boolean isDouble = value.getData().length == DoubleDataWord.BYTES;
+            byte[] data = (isDouble) ?
+                RLP.encodeElement(value.getData()) :
+                RLP.encodeElement(value.getNoLeadZeroesData());
+
+            storageTrie.update(key.getData(), data);
         }
 
         this.setDirty(true);
         this.rlpEncoded = null;
     }
 
+    /**
+     * Returns a DataWord whose 16 bytes consists of all zeros if key is not in the database.
+     * Otherwise returns the IDataWord value corresponding to key in the database.
+     *
+     * @param key The key to query.
+     * @return the corresponding value or a zero DataWord if no such value.
+     */
     @Override
-    public DataWord get(DataWord key) {
-        DataWord result = DataWord.ZERO;
+    public IDataWord get(IDataWord key) {
+        IDataWord result = DataWord.ZERO;
 
         byte[] data = storageTrie.get(key.getData());
-        if (data.length > 0) {
-            byte[] dataDecoded = RLP.decode2(data).get(0).getRLPData();
-            result = new DataWord(dataDecoded);
+        if (data.length >= DoubleDataWord.BYTES) {
```

Comment:
It is possible that rlp encoding produces an array longer than the size of a double data word, because it prepends the encoding length to the start of the array; but it cannot be shorter

Reply:
can the single `DataWord` be longer?

Reply:
yes but only one extra byte will be prepended to denote its length

Reply:
ok. looks good.